stan/math/opencl/kernel_generator/multi_result_kernel.hpp

```diff
@@ -531,11 +541,16 @@ class results_cl {
    * @param expression expression
    * @return a tuple of pair of result and expression
    */
-  template <typename T_result, typename T_expression,
-            require_all_not_t<is_without_output<T_expression>,
-                              conjunction<internal::is_scalar_check<T_result>,
-                                          std::is_arithmetic<std::decay_t<
-                                              T_expression>>>>* = nullptr>
+  template <
+      typename T_result, typename T_expression,
+      require_all_not_t<
+          is_without_output<T_expression>,
+          conjunction<internal::is_scalar_check<T_result>,
+                      std::is_arithmetic<std::decay_t<T_expression>>>,
+          conjunction<
+              is_matrix_cl<T_result>, is_matrix_cl<T_expression>,
+              std::is_same<value_type_t<T_expression>, value_type_t<T_result>>,
+              std::is_rvalue_reference<T_expression&&>>>* = nullptr>
```

Comment:
If this is strictly for rvalue references why are we forwarding in the below? And why does this one need to be an rvalue reference?

Reply:
There are two things that can be r- or l-value references. One is tuple (which in this case must be a rvalue reference and than there are tuple elements that can be either.